src/datasources/results/query-handlers/query-steps/QueryStepsDataSource.ts

```diff
@@ -0,0 +1,139 @@
+import { DataQueryRequest, DataFrameDTO, FieldType } from '@grafana/data';
+import { OutputType, QueryType } from 'datasources/results/types/types';
+import { QuerySteps, QueryStepsResponse, StepsProperties, StepsPropertiesOptions, StepsResponseProperties } from 'datasources/results/types/QuerySteps.types';
+import { ResultsDataSourceBase } from 'datasources/results/ResultsDataSourceBase';
+
+export class QueryStepsDataSource extends ResultsDataSourceBase {
+  queryStepsUrl = this.baseUrl + '/v2/query-steps';
+
+  defaultQuery = {
+    queryType: QueryType.Steps,
+    outputType: OutputType.Data,
+    properties: [
+      StepsProperties.name, 
+      StepsProperties.status, 
+      StepsProperties.totalTimeInSeconds
+    ] as StepsProperties[],
+    recordCount: 10000,
+  };
+
+  async querySteps(
+    filter?: string,
+    orderBy?: string,
+    projection?: StepsProperties[],
+    take?: number,
+    descending?: boolean,
+    returnCount = false
+  ): Promise<QueryStepsResponse> {
+    try {
+      return await this.post<QueryStepsResponse>(`${this.queryStepsUrl}`, {
+        filter,
+        orderBy,
+        descending,
+        projection,
+        take,
+        returnCount,
+      });
+    } catch (error) {
+      throw new Error(`An error occurred while querying steps: ${error}`);
+    }
+  }
+
+  async runQuery(query: QuerySteps, options: DataQueryRequest): Promise<DataFrameDTO> {
+    const projection = query.showMeasurements
+      ? [...new Set([...(query.properties || []), StepsPropertiesOptions.DATA])]
+      : query.properties;
+
+    const responseData = await this.querySteps(
+      this.getTimeRangeFilter(options, query.useTimeRange, query.useTimeRangeFor),
+      query.orderBy,
+      projection as StepsProperties[],
+      query.recordCount,
+      query.descending,
+      true
+    );
+
+    if (responseData.steps.length === 0) {
+      return {
+        refId: query.refId,
+        fields: [],
+      };
+    }
+
+    if (query.outputType === OutputType.Data) {
+      const stepsResponse = responseData.steps;
+
+      const selectedFields = (query.properties || []).filter(field => Object.keys(stepsResponse[0]).includes(field));
```

Comment:
```suggestion
   const stepResponseKeys = new Set(Object.keys(stepsResponse[0]));
    const selectedFields = (query.properties || []).filter(field => stepKeys.has(field));
```
Computing the stepsResponse keys once reduces the computation of redundant computation when filtering

Reply:
Updated